### PR TITLE
Add PR check and update syncset generation script to be consistent wi…

### PR DESCRIPTION
**Pull request by jewzaam**
…th managed-cluster-config

### CODE REVIEW

File: scripts/templates/template.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  value: REPLACED_BY_SCRIPT
+  required: true
+metadata:
+  name: selectorsyncset-template
+objects:
+# replace old SSS w/ nothing so we don't have manual cleanup to do
+- apiVersion: hive.openshift.io/v1alpha1
+  kind: SelectorSyncSet
+  metadata:
+    name: osd-${REPO_NAME}
```
Comment (lisa) [inline scripts/templates/template.yaml]: @jewzaam Similar, `osd` prefix?

Comment (jewzaam) [reply scripts/templates/template.yaml]: @lisa review full context of the object, the `resources` property is an empty array.  This is so we have a SSS deployed with the current (and now old) name that has "osd-" prefix that wipes out the resources.  We can later delete it.  We don't have a way through the template to delete the SSS.  And if we do, I'd address it in a separate PR anyway as it'll apply to managed-cluster-config and all the exporters.

File: scripts/generate_syncset.py [hunk 2]
```diff
@@ -27,33 +55,72 @@ def get_all_yaml_obj(file_paths):
...
+    o = copy.deepcopy(obj)
+    # Get all yaml files as array of yaml objects
+    yamls = get_all_yaml_obj(get_all_yaml_files(directory))
+    if len(yamls) == 0:
+        return
+
+    yamls = sorted(yamls)
+    for y in yamls:
+        if 'patch' in y:
+            if not 'patches' in o['spec']:
```
Comment (aaronlevy) [inline scripts/generate_syncset.py]: We check for `patches` and `resources` subkeys, but is it safe to assume `spec` exists?

Also, we're performing this check on every iteration of the loop, but `o` hasn't changed. Could we just check `o` outside the loop and setup necessary keys first?

File: scripts/generate_syncset.py [hunk 3]
```diff
@@ -27,33 +55,72 @@ def get_all_yaml_obj(file_paths):
...
+    for p in template_data['parameters']:
+        if p['name'] == 'REPO_NAME':
+            p['value'] = arguments.repo_name
+
+    # for each subdir of yaml_directory append 'object' to template
+    for (dirpath, dirnames, filenames) in os.walk(arguments.yaml_directory):
+        if filenames:
+            sss_name = dirpath.replace('/','-')
+            if sss_name == arguments.yaml_directory:
+                # files in the root dir, use repo-name for SSS name
```
Comment (aaronlevy) [inline scripts/generate_syncset.py]: nit: if variable is abbreviated, might be helpful to just spell out in comments. I had a moment of "what is SSS"

File: scripts/generate_syncset.py [hunk 4]
```diff
@@ -27,33 +55,72 @@ def get_all_yaml_obj(file_paths):
         objects = get_yaml_all(file)
         for obj in objects:
             yaml_objs.append(obj)
+    yaml_objs = sorted(yaml_objs)
     return yaml_objs
 
+def process_yamls(name, directory, obj):
```
Comment (aaronlevy) [inline scripts/generate_syncset.py]: nit: "process" is a bit vague here to someone unfamiliar (me). Might suggest adding a docstring with more details of the intention.

File: scripts/generate_syncset.py [hunk 5]
```diff
@@ -1,8 +1,34 @@
-#!/usr/bin/python
+#!/usr/bin/env python
```
Comment (aaronlevy) [inline scripts/generate_syncset.py]: If we don't already - it might be helpful to have input/expected output tests for this script. This is somewhat hard to parse for accuracy just from code review. And along those lines - is this shared behavior or just for this repo?

File: hack/00-osd-managed-prometheus-exporter-stuck-ebs-vols.selectorsyncset.yaml.tmpl [hunk 6]
```diff
@@ -279,7 +171,122 @@ objects:
...
+      - apiGroups:
+        - machine.openshift.io
+        resources:
+        - machines
+        verbs:
+        - get
+        - list
+    - apiVersion: v1
+      data:
+        main.py: "#!/usr/bin/python\n\nimport openshift as oc\nimport time\nimport
```
[the remark this answers is not in the document]
Comment (jewzaam) [reply hack/00-osd-managed-prometheus-exporter-stuck-ebs-vols.selectorsyncset.yaml.tmpl]: Outside of the scope of this PR but we could look at adjusting how this is generated.  I haven't looked at that code to be honest.

File: hack/app_sre_pr_check.sh [hunk 7]
```diff
@@ -2,5 +2,17 @@
 
 set -exv
 
+# if running `make` changes anything, fail the build
+# order is inconsistent across systems, sort the template file.. it's not perfect but it's better than nothing
+cat hack/00-osd-*.selectorsyncset.yaml.tmpl | sort > sorted-before.yaml.tmpl
```
[the remark this answers is not in the document]
Comment (jewzaam) [reply hack/app_sre_pr_check.sh]: It's possible we're at a point the sort can be removed.  Originally the yaml output was ordered inconsistently.  Agree this isn't great as it is, as noted in the comment.  But it's better than nothing and catches course grained changes not being applied.

The template is committed in the repo, git could be used to check diff if running make doesn't reorder the yaml properties.  I'd rather tackle that in a separate PR if we can.

File: hack/00-osd-managed-prometheus-exporter-stuck-ebs-vols.selectorsyncset.yaml.tmpl (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaronlevy) [reply hack/00-osd-managed-prometheus-exporter-stuck-ebs-vols.selectorsyncset.yaml.tmpl]: ack

File: hack/app_sre_pr_check.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaronlevy) [reply hack/app_sre_pr_check.sh]: That's fine. Not sure how we're tracking cleanup tasks like this, but something along the lines of "ensure yaml output is consistent + use git for diff check" might be good to backlog. Not high priority - but could be a good intro / on-boarding task if we're adding new people to the team.

File: hack/00-osd-managed-prometheus-exporter-stuck-ebs-vols.selectorsyncset.yaml.tmpl [hunk 8]
```diff
@@ -6,11 +6,21 @@ objects:
 - apiVersion: hive.openshift.io/v1alpha1
   kind: SelectorSyncSet
   metadata:
-    generation: 1
+    name: osd-${REPO_NAME}
```
[the remark this answers is not in the document]
Comment (jewzaam) [reply hack/00-osd-managed-prometheus-exporter-stuck-ebs-vols.selectorsyncset.yaml.tmpl]: Look at the rest of the SSS and it's an empty resources array.  This remains simply to deploy an empty SSS replacing the one with content we already have.

File: Makefile [hunk 9]
```diff
@@ -78,8 +84,8 @@ deploy/060_servicemonitor.yaml: resources/060_servicemonitor.yaml.tmpl
 	@$(call generate_file,060_servicemonitor)
 
 .PHONY: generate-syncset
-generate-syncset: 
-	scripts/generate_syncset.py -t ${SELECTOR_SYNC_SET_TEMPLATE} -y ${YAML_DIRECTORY} -d ${SELECTOR_SYNC_SET_DESTINATION} -c ${GIT_HASH}
+generate-syncset:
+	docker run --rm -v `pwd`:`pwd` python:2.7.15 /bin/sh -c "cd `pwd`; pip install pyyaml; scripts/generate_syncset.py -t ${SELECTOR_SYNC_SET_TEMPLATE_DIR} -y ${YAML_DIRECTORY} -d ${SELECTOR_SYNC_SET_DESTINATION} -r ${REPO_NAME}"
```
Comment (lisa) [inline Makefile]:
```suggestion
	docker run --rm -v `pwd -P`:`pwd -P` python:2.7.15 /bin/sh -c "cd `pwd`; pip install pyyaml; scripts/generate_syncset.py -t ${SELECTOR_SYNC_SET_TEMPLATE_DIR} -y ${YAML_DIRECTORY} -d ${SELECTOR_SYNC_SET_DESTINATION} -r ${REPO_NAME}"
```
